modules/navier_stokes/src/userobjects/RhieChowMassFlux.C

```diff
@@ -0,0 +1,516 @@
+
+    // On internal face we do a regular interpolation with geometric weights
+    if (_vel[0]->isInternalFace(*fi))
+    {
+      const auto & elem_info = *fi->elemInfo();
+      const auto & neighbor_info = *fi->neighborInfo();
+
+      Real elem_rho = _rho(makeElemArg(fi->elemPtr()), time_arg);
+      Real neighbor_rho = _rho(makeElemArg(fi->neighborPtr()), time_arg);
+
+      for (const auto dim_i : index_range(_vel))
+        interpolate(InterpMethod::Average,
+                    density_times_velocity(dim_i),
+                    _vel[dim_i]->getElemValue(elem_info, time_arg) * elem_rho,
+                    _vel[dim_i]->getElemValue(neighbor_info, time_arg) * neighbor_rho,
+                    *fi,
+                    true);
+    }
+    // On the boundary, we just take the boundary values
+    else
+    {
+      const Elem * const boundary_elem =
+          hasBlocks(fi->elemPtr()->subdomain_id()) ? fi->elemPtr() : fi->neighborPtr();
+
+      const Moose::FaceArg boundary_face{
+          fi, Moose::FV::LimiterType::CentralDifference, true, false, boundary_elem};
+
+      const Real face_rho = _rho(boundary_face, time_arg);
+      for (const auto dim_i : index_range(_vel))
+        density_times_velocity(dim_i) =
+            face_rho * raw_value((*_vel[dim_i])(boundary_face, time_arg));
+    }
+
+    _face_mass_flux[fi->id()] = density_times_velocity * fi->normal();
+  }
+}
+
+Real
+RhieChowMassFlux::getMassFlux(const FaceInfo & fi) const
+{
+  return _face_mass_flux.evaluate(&fi);
+}
+
+void
+RhieChowMassFlux::computeFaceMassFlux()
+{
+  using namespace Moose::FV;
+
+  const auto time_arg = Moose::currentState();
+
+  PetscVectorReader p_reader(*_p->sys().system().current_local_solution);
+  for (auto & fi : _fe_problem.mesh().faceInfo())
+  {
+    _p_diffusion_kernel->setCurrentFaceInfo(fi);
+    _p_diffusion_kernel->setCurrentFaceArea(1.0);
+
+    Real p_grad_flux = 0.0;
+    if (_p->isInternalFace(*fi))
+    {
+      const auto & elem_info = *fi->elemInfo();
+      const auto & neighbor_info = *fi->neighborInfo();
+      const auto elem_dof = elem_info.dofIndices()[_global_pressure_system_number][0];
+      const auto neighbor_dof = neighbor_info.dofIndices()[_global_pressure_system_number][0];
+
+      const auto p_elem_value = p_reader(elem_dof);
+      const auto p_neighbor_value = p_reader(neighbor_dof);
+
+      const auto elem_matrix_contribution = _p_diffusion_kernel->computeElemMatrixContribution();
+      const auto neighbor_matrix_contribution =
+          _p_diffusion_kernel->computeNeighborMatrixContribution();
+      const auto elem_rhs_contribution =
+          _p_diffusion_kernel->computeElemRightHandSideContribution();
+
+      p_grad_flux = (p_neighbor_value * neighbor_matrix_contribution +
+                     p_elem_value * elem_matrix_contribution) -
+                    elem_rhs_contribution;
+    }
+    else if (auto * bc_pointer = _p->getBoundaryCondition(*fi->boundaryIDs().begin()))
```

Comment:
could this be tricky if fi has multiple boundary ids but the bc is only associated with the second one?

Reply:
Added an assert to ensure one boundary ID per face.